### PR TITLE
perf: avoid cloning HashBuilder input

### LINKED ISSUE #45: perf: store `HashBuilder.value` as a `Vec<u8>` and track kind separately
The current API of `HashBuilder::leaf` creates a new allocation for each new value. This is wasteful, and should instead re-use an internal buffer like we're doing for RLP.

Practically, this means replacing `value` with a `Vec<u8>` and a separate flag to track whether it is a hash or a leaf value. The `add_*` APIs should remain the same, and internally they will clear the buffer, copy the new value in, and set the flag.

### PR DESCRIPTION
**Pull request by DaniPopes**
Closes https://github.com/alloy-rs/trie/issues/45

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
makes sense

File: src/hash_builder/input.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        self.kind = HashBuilderInputKind::default();
+    }
+}
+
+/// The kind of the current hash builder input.
+#[doc(hidden)] // Not public API.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "arbitrary", derive(derive_arbitrary::Arbitrary, proptest_derive::Arbitrary))]
+#[allow(unreachable_pub)]
```
Comment (rkrasiuk) [inline src/hash_builder/input.rs]:
```suggestion
```

File: src/hash_builder/input.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+pub enum HashBuilderInputKind {
```
Comment (rkrasiuk) [inline src/hash_builder/input.rs]: why the input rename? nodes have values, not inputs

File: src/hash_builder/mod.rs [hunk 3]
```diff
@@ -154,11 +154,20 @@ impl HashBuilder {
         root
     }
 
-    fn set_key_value<T: Into<HashBuilderValue>>(&mut self, key: Nibbles, value: T) {
-        trace!(target: "trie::hash_builder", key = ?self.key, value = ?self.value, "old key/value");
+    #[inline]
+    fn set_input(&mut self, key: Nibbles, input: HashBuilderInputRef<'_>) {
```
Comment (rkrasiuk) [inline src/hash_builder/mod.rs]: same re rename

File: src/hash_builder/input.rs (hunk 1 above)
Comment (DaniPopes) [reply src/hash_builder/input.rs]: doesnt need to be pub, remainder from previous changes

File: src/hash_builder/input.rs (hunk 2 above)
Comment (DaniPopes) [reply src/hash_builder/input.rs]: was thinking to include key or something i forgot, no reason now

File: src/hash_builder/input.rs (hunk 1 above)
Comment (DaniPopes) [reply src/hash_builder/input.rs]: nvm it does because of proptest derive, just made it pub officially

**Pull request merged**